WASI P3 C bindings: 'assigning from compatible type' error

Using a pre-release build of wit-bindgen-cli: 0.57.1 (d63511735 2026-05-27). WASI SDK 33.  Repro code for this is in https://github.com/itowlson/c-async-fear-fear-fear.

I am running `wit-bindgen c` over the Spin 4 WITs.  This works without errors.  But when I try to build the generated code, I get the following errors:

```
$ $WASI_SDK_PATH/bin/wasm32-wasip2-clang --sysroot=$WASI_SDK_PATH/share/wasi-sysroot test.c bindings/http_trigger.c bindings/http_trigger_component_type.o -mexec-model=reactor -o test.wasm

bindings/http_trigger.c:13222:46: error: incompatible pointer types initializing 'wasi_http_0_2_6_types_field_value_t *'
      (aka 'struct wasi_http_0_2_6_types_field_value_t *') with an expression of type
      'wasi_http_0_3_0_rc_2026_03_15_types_field_value_t *' (aka 'struct
      wasi_http_0_3_0_rc_2026_03_15_types_field_value_t *') [-Wincompatible-pointer-types]
 13222 |   ...{ (wasi_http_0_3_0_rc_2026_03_15_types_field_value_t*)(*((uint8_t **) (ptr + 0))), (*((size_t*) (ptr + ...
       |        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
bindings/http_trigger.c:13355:59: error: incompatible pointer types initializing 'wasi_http_0_2_6_types_field_value_t *'
      (aka 'struct wasi_http_0_2_6_types_field_value_t *') with an expression of type
      'wasi_http_0_3_0_rc_2026_03_15_types_field_value_t *' (aka 'struct
      wasi_http_0_3_0_rc_2026_03_15_types_field_value_t *') [-Wincompatible-pointer-types]
 13355 |   ...{ (wasi_http_0_3_0_rc_2026_03_15_types_field_value_t*)(*((uint8_t **) (ptr + sizeof(void*)))), (*((size...
       |        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
bindings/http_trigger.c:22423:46: error: incompatible pointer types initializing 'wasi_http_0_2_6_types_field_value_t *'
      (aka 'struct wasi_http_0_2_6_types_field_value_t *') with an expression of type
      'wasi_http_0_2_0_types_field_value_t *' (aka 'struct wasi_http_0_2_0_types_field_value_t *')
      [-Wincompatible-pointer-types]
 22423 |   ...(http_trigger_list_field_value_t) { (wasi_http_0_2_0_types_field_value_t*)(*((uint8_t **) (ptr + 0))), ...
       |                                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
bindings/http_trigger.c:26226:22: error: assigning to 'spin_redis_redis_payload_t' (aka
      'struct spin_redis_redis_payload_t') from incompatible type 'fermyon_spin_2_0_0_redis_payload_t' (aka 'struct
      fermyon_spin_2_0_0_redis_payload_t')
 26226 |   ...= (fermyon_spin_2_0_0_redis_payload_t) { (uint8_t*)(*((uint8_t **) (ptr + (2*sizeof(void*))))), (*((size_t*) (ptr + (3*sizeof(void*))))) };
       |      ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
4 errors generated.
```

The Spin WITs do contain P2 and P3 versions of interfaces, so it's expected that there are multiple types representing e.g. "Redis payload" or "HTTP fields collection", but as far as I know this is allowed, and the Rust bindings generator makes no objection.  Then again, there are many many other APIs in these WITs that have P2 and P3 versions, and they're not appearing in the error list, so I am not sure if this is relevant or a red herring!

--- a/crates/c/src/lib.rs
+++ b/crates/c/src/lib.rs
@@ -1068,6 +1068,17 @@ fn is_prim_type(resolve: &Resolve, ty: &Type) -> bool {
 }
 
 fn is_prim_type_id(resolve: &Resolve, id: TypeId) -> bool {
+    // A named type contributes its (interface-local) name to the generated
+    // type name rather than its structure, so even when its underlying
+    // representation is primitive it must not be treated as a world-shareable
+    // "primitive" type. Otherwise two interfaces that each define a
+    // differently-typed but identically-named type (e.g. `field-value` in two
+    // versions of `wasi:http/types`) collapse onto the same anonymous type
+    // name, producing structs with incompatible element pointer types.
+    // See https://github.com/bytecodealliance/wit-bindgen/issues/1621.
+    if resolve.types[id].name.is_some() {
+        return false;
+    }
     match &resolve.types[id].kind {
         TypeDefKind::List(elem) => is_prim_type(resolve, elem),
 

--- a/tests/codegen/issue-1621.wit
+++ b/tests/codegen/issue-1621.wit
@@ -1,0 +1,16 @@
+package my:test;
+
+interface a {
+  type t = list<u8>;
+  f: func() -> list<t>;
+}
+
+interface b {
+  type t = list<u8>;
+  g: func() -> list<t>;
+}
+
+world w {
+  import a;
+  import b;
+}

--- a/tests/runtime/flavorful/runner.c
+++ b/tests/runtime/flavorful/runner.c
@@ -79,7 +79,7 @@ void exports_runner_run() {
     test_list_typedef3_t b;
     b.ptr = &b_str;
     b.len = 1;
-    runner_tuple2_list_typedef2_list_typedef3_t ret;
+    test_tuple2_list_typedef2_list_typedef3_t ret;
     test_list_typedefs(&a, &b, &ret);
 
     assert(memcmp(ret.f0.ptr, "typedef3", ret.f0.len) == 0);

--- a/tests/runtime/flavorful/test.c
+++ b/tests/runtime/flavorful/test.c
@@ -65,7 +65,7 @@ bool exports_test_errno_result(exports_test_my_errno_t *err) {
 void exports_test_list_typedefs(
     exports_test_list_typedef_t *a,
     exports_test_list_typedef3_t *c,
-    test_tuple2_list_typedef2_list_typedef3_t *ret) {
+    exports_test_tuple2_list_typedef2_list_typedef3_t *ret) {
   assert(memcmp(a->ptr, "typedef1", a->len) == 0);
   exports_test_list_typedef_free(a);
 

